File: codigo/java/BuscaBinariaCasos.java
package codigo.java;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class BuscaBinariaCasos {

    static final int RUNS = 50;

    // -------- Função de busca binária --------
    public static int buscaBinaria(int[] vetor, int chave) {
        int esquerda = 0, direita = vetor.length - 1;
        while (esquerda <= direita) {
            int meio = (esquerda + direita) / 2;
            if (vetor[meio] == chave)
                return meio;
            else if (vetor[meio] < chave)
                esquerda = meio + 1;
            else
                direita = meio - 1;
        }
        return -1;
    }

    // -------- Leitura do vetor CSV --------
    public static int[] lerVetor(String caminho, int n) {
        int[] vetor = new int[n];
        try (BufferedReader br = new BufferedReader(new FileReader(caminho))) {
            String linha = br.readLine();
            if (linha == null) return null;
            String[] valores = linha.split(",");
            for (int i = 0; i < n; i++) {
                vetor[i] = Integer.parseInt(valores[i].trim());
            }
        } catch (IOException e) {
            System.out.println("Erro ao abrir " + caminho);
            return null;
        }
        return vetor;
    }

    // -------- Medição do tempo em nanossegundos --------
    public static double medirTempoExecucaoNs(int[] vetor, int chave) {
        long inicio = System.nanoTime();
        buscaBinaria(vetor, chave);
        long fim = System.nanoTime();
        return (double) (fim - inicio);
    }

    // -------- Cálculos estatísticos --------
    public static double calcularMedia(double[] tempos) {
        double soma = 0.0;
        for (double t : tempos) soma += t;
        return soma / tempos.length;
    }

    public static double calcularDesvioPadrao(double[] tempos, double media) {
        double soma = 0.0;
        for (double t : tempos) {
            double dif = t - media;
            soma += dif * dif;
        }
        return Math.sqrt(soma / tempos.length);
    }

    // -------- Função principal --------
    public static void main(String[] args) {
        int[] tamanhos = {10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000, 90000, 100000};
        String[] casos = {"MC", "CM", "PC"};
        Random rand = new Random();

        // Diretórios e arquivos
        Path dirResultados = Paths.get("resultados");
        Path dirEstatisticas = dirResultados.resolve("estatisticas");

        try {
            Files.createDirectories(dirEstatisticas);
        } catch (IOException e) {
            System.out.println("Erro ao criar diretórios");
            return;
        }

        File arquivoEst = dirEstatisticas.resolve("resultados_casos_java.csv").toFile();

        try (PrintWriter saidaEst = new PrintWriter(new FileWriter(arquivoEst))) {

            saidaEst.println("n,caso,media_ns,desvio_ns");
            //saidaBrutos.println("n,caso,run,tempo_ns");

            for (int n : tamanhos) {

                for (int tipo = 0; tipo < 3; tipo++) {
                    double[] tempos = new double[RUNS];

                    for (int run = 1; run <= RUNS; run++) {
                        String caminho = String.format("dados\\n%06d\\run_%03d.csv", n, run);
                        int[] vetor = lerVetor(caminho, n);
                        if (vetor == null) continue;

                        int chave;
                        if (tipo == 0) { // MC - melhor caso
                            chave = vetor[n / 2];
                        } else if (tipo == 1) { // CM - caso médio
                            chave = rand.nextBoolean() ? vetor[rand.nextInt(n)] : vetor[n - 1] + rand.nextInt(500);
                        } else { // PC - pior caso
                            chave = vetor[n - 1] + 10000; // garante inexistente
                        }

                        double tempo = medirTempoExecucaoNs(vetor, chave);
                        tempos[run - 1] = tempo;

                    }

                    double media = calcularMedia(tempos);
                    double desvio = calcularDesvioPadrao(tempos, media);

                    saidaEst.printf("%d,%s,%.2f,%.2f%n", n, casos[tipo], media, desvio);
                    System.out.printf("n=%d (%s) concluído (média=%.2f ns, desvio=%.2f ns)%n", n, casos[tipo], media, desvio);
                }
            }

        
        } catch (IOException e) {
            System.out.println("Erro ao criar arquivos de saída.");
        }
    }
}

File: codigo/python/busca_binaria_casos.py
import os
import csv
import math
import time
import random

RUNS = 50

# -------- Função de busca binária --------
def busca_binaria_casos(vetor, chave):
    esquerda, direita = 0, len(vetor) - 1
    while esquerda <= direita:
        meio = (esquerda + direita) // 2
        if vetor[meio] == chave:
            return meio
        elif vetor[meio] < chave:
            esquerda = meio + 1
        else:
            direita = meio - 1
    return -1

# -------- Leitura do vetor CSV --------
def ler_vetor(caminho):
    try:
        with open(caminho, newline='') as f:
            reader = csv.reader(f)
            linha = next(reader)
            vetor = [int(x) for x in linha]
        return vetor
    except Exception as e:
        print(f"Erro ao abrir {caminho}: {e}")
        return None

# -------- Medição do tempo em nanossegundos --------
def medir_tempo_execucao_ns(vetor, chave):
    inicio = time.perf_counter_ns()
    busca_binaria_casos(vetor, chave)
    fim = time.perf_counter_ns()
    return fim - inicio

# -------- Cálculos estatísticos --------
def calcular_media(valores):
    return sum(valores) / len(valores)

def calcular_desvio_padrao(valores, media):
    soma = sum((v - media) ** 2 for v in valores)
    return math.sqrt(soma / len(valores))

# -------- Função principal --------
def main():
    tamanhos = [10000, 20000, 30000, 40000, 50000,
                60000, 70000, 80000, 90000, 100000]
    casos = ["MC", "CM", "PC"]

    # Cria os diretórios de saída, se não existirem
    os.makedirs("resultados/estatisticas", exist_ok=True)
    os.makedirs("resultados/brutos", exist_ok=True)

    arquivo_est = "resultados/estatisticas/resultados_casos_python.csv"
    arquivo_brutos = "resultados/brutos/resultados_python_run.csv"

    with open(arquivo_est, "w", newline='') as f_est, \
         open(arquivo_brutos, "w", newline='') as f_brutos:

        writer_est = csv.writer(f_est)
        writer_brutos = csv.writer(f_brutos)

        writer_est.writerow(["n", "caso", "media_ns", "desvio_ns"])
        writer_brutos.writerow(["n", "caso", "run", "tempo_ns"])

        for n in tamanhos:
            for tipo, caso in enumerate(casos):
                tempos = []

                for run in range(1, RUNS + 1):
                    caminho = f"dados/n{n:06d}/run_{run:03d}.csv"
                    vetor = ler_vetor(caminho)
                    if vetor is None:
                        continue

                    if tipo == 0:  # MC - melhor caso
                        chave = vetor[len(vetor) // 2]
                    elif tipo == 1:  # CM - caso médio
                        if random.random() < 0.5:
                            chave = random.choice(vetor)
                        else:
                            chave = vetor[-1] + random.randint(1, 500)
                    else:  # PC - pior caso
                        chave = vetor[-1] + 10000

                    tempo = medir_tempo_execucao_ns(vetor, chave)
                    tempos.append(tempo)
                    writer_brutos.writerow([n, caso, run, tempo])

                if tempos:
                    media = calcular_media(tempos)
                    desvio = calcular_desvio_padrao(tempos, media)
                    writer_est.writerow([n, caso, media, desvio])
                    print(f"n={n} ({caso}) concluído (média={media:.2f} ns, desvio={desvio:.2f} ns)")

if __name__ == "__main__":
    main()
